parastell/invessel_build.py

```diff
@@ -256,6 +302,150 @@ def generate_components(self):
+                    corner3 = next_rib.mb_verts[rib_pt_index + 1]
+                    corner4 = next_rib.mb_verts[rib_pt_index]
+                    corners = [corner1, corner2, corner3, corner4]
+                    mb_tris += create_moab_tris_from_verts(
+                        corners,
+                        self.dag_model.mb,
+                        reverse=True if i == 0 else False,
+                    )
+            dagmc_surface = dagmc.Surface.create(self.dag_model)
+            self.dag_model.mb.add_entities(dagmc_surface.handle, mb_tris)
+            curved_surface_ids.append(dagmc_surface.id)
+        self.curved_surface_ids = curved_surface_ids
+
+    def generate_end_cap_surfaces_pydagmc(self):
+        """Generate the faceted representation of the planar end cap surfaces
+        and add them to the PyDAGMC model, remembering the surface ids."""
+        end_cap_surface_ids = []
+        for surface, next_surface in zip(
+            list(self.Surfaces.values())[0:-1],
+            list(self.Surfaces.values())[1:],
+        ):
+            mb_tris = self.connect_ribs_with_tris_moab(
+                surface.Ribs[0], next_surface.Ribs[0]
+            )
+            end_cap_start = dagmc.Surface.create(self.dag_model)
+            self.mbc.add_entities(end_cap_start.handle, mb_tris)
+
+            mb_tris = self.connect_ribs_with_tris_moab(
+                surface.Ribs[-1], next_surface.Ribs[-1], reverse=True
+            )
+            end_cap_end = dagmc.Surface.create(self.dag_model)
+            self.mbc.add_entities(end_cap_end.handle, mb_tris)
+            end_cap_surface_ids.append(
+                list(self.dag_model.surfaces_by_id.keys())[-2:]
+            )
+        self.end_cap_surface_ids = end_cap_surface_ids
+
+    def generate_volumes_pydagmc(self):
+        """Use the curved surface and end cap surface IDs to build the
+        the volumes by applying the correct surface sense to each surface."""
+        # make the volumes and apply surface sense
+        for i in range(len(self.Surfaces) - 1):
+            dagmc.Volume.create(self.dag_model)
+
+        for i, surface_id in enumerate(self.curved_surface_ids):
+            # if it is the first surface, it goes to the implicit complement
+            if i == 0:
+                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
+                    self.dag_model.volumes_by_id[i + 1],
+                    None,
+                ]
+            elif i != len(self.curved_surface_ids) - 1:
+                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
+                    self.dag_model.volumes_by_id[i],
+                    self.dag_model.volumes_by_id[i + 1],
+                ]
+            # if it the last surface it goes to the implicit complement
+            else:
+                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
+                    self.dag_model.volumes_by_id[i],
+                    None,
+                ]
```

Comment:
Not critical, but I think it may make more sense logically/be more readable to switch these `elif` and `else` statements, instead using
```suggestion
            # if it the last surface it goes to the implicit complement
            elif i != len(self.curved_surface_ids) - 1:
                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
                    self.dag_model.volumes_by_id[i],
                    None,
                ]
            else:
                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
                    self.dag_model.volumes_by_id[i],
                    self.dag_model.volumes_by_id[i + 1],
                ]
```

Reply:
Ope, forgot to update the condition - good catch!

Reply:
Going with @gonuke's suggestion since I also like keeping them in order from inside out.